src/migrator/async.ts

```diff
@@ -0,0 +1,301 @@
+					await lockMigrations(tx, modelName, async () => {
+						const migrationState = await readMigrationStatus(tx, key);
+
+						if (!migrationState) {
+							// migration status is unclear stop the migrator
+							// or migration should stop
+							(sbvrUtils.api.migrations?.logger.info ?? console.info)(
+								`stopping async migration: ${key}`,
+							);
+							clearInterval(asyncScheduler);
+							return;
+						}
+						try {
+							// clear the interval to avoid re-triggering before finishing this migration query
+							clearInterval(asyncScheduler);
+
+							// sync on the last execution time between instances
+							// precondition: All running instances are running on the same time/block
+							// skip execution
+							if (migrationState.last_run_time) {
+								const durationSinceLastRun =
+									Date.now() - migrationState.last_run_time.getTime();
+								if (
+									(migrationState.is_backoff &&
+										durationSinceLastRun < initMigrationState.backoffDelayMS) ||
+									(!migrationState.is_backoff &&
+										durationSinceLastRun < initMigrationState.delayMS)
+								) {
+									// will still execute finally block where the migration lock is released.
+									return;
+								}
+							}
+							// set last run time and run counter only when backoff time sync between
+							// competing instances is in sync
+							migrationState.last_run_time = new Date();
+							migrationState.run_counter += 1;
+
+							let migratedRows = 0;
+							// here a separate transaction is needed as this migration may fail
+							// when it fails it would break the transaction for managing the migration status
+							await sbvrUtils.db.transaction(async (migrationTx) => {
+								migratedRows =
+									(await asyncRunnerMigratorFn?.(migrationTx)) || 0;
+							});
+
+							migrationState.migrated_rows += migratedRows;
+							if (migratedRows === 0) {
+								// when all rows have been catched up once we only catch up less frequently
+								migrationState.is_backoff = true;
+								if (!migrationState.converged_time) {
+									// only store the first time when migrator converged to all data migrated
+									migrationState.converged_time = new Date();
+								}
+							} else {
+								// Only here for the case that after backoff more rows need to be caught up faster
+								// If rows have been updated recently we start the interval again with normal frequency
+								migrationState.is_backoff = false;
+							}
+						} catch (err: unknown) {
+							migrationState.error_counter++;
+							if (
+								err instanceof Error &&
+								migrationState.error_counter %
+									initMigrationState.error_threshold ===
+									0
+							) {
+								migrationState.last_error_message = `${err.name} ${err.message}`;
+								(sbvrUtils.api.migrations?.logger.error ?? console.error)(
+									`${key}: ${err.name} ${err.message}`,
+								);
+								migrationState.is_backoff = true;
+							}
```

Comment:
We probably still want to log/do something if `err` is not an `instance of Error` because that'll be pretty surprising/bad

Reply:
Yes, that's right. Thanks!
Adding logging for key (migration key) : {err}